### ISSUE TITLE
Morphological object classification along with instance segmentation

### ISSUE THREAD

**Issue by Aseem139** [BiaPyX/BiaPy#135]
Hi Daniel,

I have recently started using BiaPy and find it really convenient to use. Many congratulations on your Nature Methods papers.I am mainly interested in instance segmentation. I have so far tried the nuclei dataset and I am attaching the config file I used and the training results.

Config File:

[3d_instance_segmentation.yaml.txt](https://github.com/user-attachments/files/21135348/3d_instance_segmentation.yaml.txt)

Training Results: 

<img width="1915" height="518" alt="Image" src="https://github.com/user-attachments/assets/699c8b99-20d0-40d4-bbf8-380359ad7e51" />

<img width="1470" height="370" alt="Image" src="https://github.com/user-attachments/assets/6a1efc2b-bdec-4885-b68f-1873e56ce2e0" />

I have a few questions:

1) The IOU values for the contour channel don't look all that good even though the overall predictions were good. Is this normal ? Should I result the relative weight of the contour channel ? 

2) For the segmentation of mitochondria that I am currently working on, I am interested in trying the classification of 3d objects in addition to instance segmentation. I understand how the training dataset must now have 2 channels, one for instance label and other for class labels. Is there any other changes I must make to my config file to tell the model to do classification along with segmentation ?

3) Are there any other changes you'd suggest in my config file that would improve segmentation quality ?

Thanks in advance for the help.

**Comment by iarganda** [comment]
Hello @Aseem139!

Thanks a lot for your message. We're glad you like BiaPy!

From your plots, it seems your training is highly overfitting. Which dataset are you using? The [3D demo from the StarDist 0.3.0 release](https://biapy.readthedocs.io/en/latest/workflows/instance_segmentation.html#example-datasets)?

**Comment by iarganda** [comment]
I see! Those are very good results indeed. Let me answer your questions:

> The IOU values for the contour channel don't look all that good even though the overall predictions were good. Is this normal ? Should I result the relative weight of the contour channel ?

This is normal. The IoU of the B channel (binary masks) is high (~0.82 in test), which means the masks of your objects are well identified. The IoU of the C channel (contour maks) is low (~0.10 in test) but this is expected since it measures the overlap between object contours in the prediction and the ground truth, and they are very thin. Nevertheless, those channels are successfully post-processed to obtain individual instance masks (see how high your F1 values are for IoUs 0.3 and 0.5). Summarizing, I would not worry at all about the low metric value of the C channel.

> For the segmentation of mitochondria that I am currently working on, I am interested in trying the classification of 3d objects in addition to instance segmentation. I understand how the training dataset must now have 2 channels, one for instance label and other for class labels. Is there any other changes I must make to my config file to tell the model to do classification along with segmentation ?

Maybe @danifranco can tell us more about it, but I believe you only need to do that and also set the variable `N_CLASSES` (under `MODEL`) to your number of classes (including the background class, that should be 0) in your yaml file.

> Are there any other changes you'd suggest in my config file that would improve segmentation quality ?

Your metrics are quite high already, is there anything you would like to improve from your results? Border quality?
